admin/class-smaily-for-wp-admin.php

```diff
@@ -105,216 +112,276 @@ public function smaily_subscription_widget_init() {
+			} elseif ( $code === 404 ) {
+				// If wrong subdomain.
+				return array(
+					'message' => __( 'Error in subdomain', 'smaily-for-wp' ),
 				);
-				break;
+			} elseif ( array_key_exists( 'error', $rqst ) ) {
+				// If there is WordPress error message.
+				return array(
+					'message' => $rqst['error'],
+				);
+			}
+			// If not determined error.
+			return array(
+				'message' => __( 'Something went wrong with request to Smaily', 'smaily-for-wp' ),
+			);
 		}
+		// Insert item to database.
+		$this->option_handler->update_api_credentials( $params );
+
+		// Return result.
+		return array(
+			'error'   => false,
+			'message' => __( 'Credentials validated.', 'smaily-for-wp' ),
+		);
+	}
 
-		// Send refresh form content (if requested).
-		if ( $refresh ) {
-			// Generate form contents.
-			$template = new Smaily_For_WP_Template( 'admin/partials/smaily-for-wp-admin-form.php' );
+	/**
+	 * Function is run when user removes saved API key.
+	 *
+	 * @since  3.0.0
+	 * @access private
+	 * @return array Response of operation.
+	 */
+	private function remove_api_key() {
+		// Delete contents of config.
+		$this->option_handler->update_api_credentials( array() );
+
+		// Set result.
+		return array(
+			'error'   => false,
+			'message' => __( 'Credentials removed.', 'smaily-for-wp' ),
+		);
+	}
+
+	/**
+	 * Function is run when user regenerates signup form.
+	 *
+	 * @since  3.0.0
+	 * @access private
+	 * @return array Response of operation.
+	 */
+	private function reset_form() {
+		$subdomain = $this->option_handler->get_api_credentials()['subdomain'];
+		$template  = $this->generate_signup_template( 'advanced.php', $subdomain );
+
+		return array(
+			'message' => __( 'Newsletter subscription form reset to default.', 'smaily-for-wp' ),
+			'error'   => false,
+			'content' => $template->render(),
+		);
+	}
 
+	/**
+	 * Function is run when user presses save button.
+	 *
+	 * @since  3.0.0
+	 * @access private
+	 * @param  array $form_data Posted form data (deserialized).
+	 * @return array Response of operation.
+	 */
+	private function save( $form_data ) {
+		// Get parameters.
+		$is_advanced = ( isset( $form_data['is_advanced'] ) && ! empty( $form_data['is_advanced'] ) ) ? '1' : '0';
+		$form        = ( isset( $form_data['form'] ) && is_string( $form_data['form'] ) ) ? $form_data['form'] : '';
+
+		// Generate new form (if empty).
+		if ( empty( $form ) ) {
 			// Load configuration data.
-			$table_name = esc_sql( $wpdb->prefix . 'smaily_config' );
-			$data       = $wpdb->get_row( "SELECT * FROM `$table_name` LIMIT 1" );
-			$template->assign( (array) $data );
+			$subdomain = $this->option_handler->get_api_credentials()['subdomain'];
+			$form      = $this->option_handler->get_form_options()['form'];
 
+			$template = $this->generate_signup_template( 'advanced.php', $subdomain, $form );
 			// Render template.
-			$result['content'] = $template->render();
+			$form = ltrim( $template->render() );
```

Comment:
It wouldn't hurt removing trailing whitespace also.

Reply:
50d9518